Feature update - surface filter

NocturnalScream opened the pull request:
- Added an option to specify the surfaces concreep should be running on (by name)
- added various options to be changed at runtime, i.e. update rate, speed etc.

NocturnalScream committed:
Feature update - surface filter

- Added an option to specify the surfaces concreep should be running on (by name)
- added various options to be changed at runtime, i.e. update rate, speed etc.

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -1,14 +1,16 @@
-MINIMUM_ROBOTS = 30
-MINIMUM_ITEM_COUNT = 200
+SURFACES = settings.global["surface concreep is running on"].value
```
Probably should rename all these now-tunable constants to be DEFAULT_<name> and then everywhere where we use them, use them directly from the `settings.global[<name>]` instead.

we should move the constants to the top of the settings.lua file so that they're easier to change quickly (instead of scrolling through the big settings dict?)

NocturnalScream replied:
> Probably should rename all these now-tunable constants to be DEFAULT_ and then everywhere where we use them, use them directly from the settings.global[<name>] instead.

Done.

> we should move the constants to the top of the settings.lua file so that they're easier to change quickly (instead of scrolling through the big settings dict?)

Not quite sure what you mean by that - one downside of all these customizable settings is that its going to be cluttered - I wish there was a way to only show "expert options" if the user wants it. I guess some options don't need to be customizable such as number of creepers per update - update interval would suffice for the most use cases?

MarximusMaximus replied:
Just to move the default constants to the top of the settings.lua file instead of here, so that they can be used in the settings file to set the defaults there (as well as instead of having to scroll through the data struct to change them).

As for lots of options, I don't think that's really a big deal. Plenty of "simpler" addons have long lists of options. :-/

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -64,15 +66,20 @@ function init()
-		end
-	end
-end
+	   for str in SURFACES:gmatch("([^,]+)") do
```
Seems to be some weird formatting here. On my local version, all indents are 4 spaces wide, but github seems to be showing tab characters for all of them? And then kind of mixed in this section?

actually... i think we should just remove this block from init, so that init is able to be called again properly and so that save startup time is lessened (if there's lots of surfaces with LOTS of roboports, this block could actually make map loads take a while)

NocturnalScream replied:
Fixed all the inconsistent formatting I could find - about removing that section - If this block does not exist, reinit will be called automatically because there are no roboports in the creeper list - I guess the benefit is that we can use init after configuration changes just to reset variables?

MarximusMaximus replied:
Yep, and the reinit running right away is fine, because it kickstarts things in a UPS friendly way, as opposed to how the init was doing it. Haven't decided what to rename reinit to though.

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -64,15 +66,20 @@ function init()
+end	
 
 -- a fake coroutine
 function reinit()
```
Thinking about renaming this to be something else, so that a true reinit function can be created. Alternative is to "fix up" init so that it can be called a second time properly.

NocturnalScream replied:
I would say we rather fix up init - reinit seems to be in a somewhat good shape if init can be fixed - so lets just remove the block (which also in reality does cause a significant amount of delay when changing settings (about 2 -4 second freeze in my savegame).

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -144,7 +151,11 @@ function _reinit_reentrant_level1_start()
+	name = str:gsub("%s+", "")
+    local surf = game.surfaces[name]
+	end
+	global.reinit_surfaces = array_shallowcopy(surf)
```
surf here will always be just a single surface? need to have it build an array instead? also, it's sort of bad form to declare a local inside a for loop anyway (need to declare it outside to make it a proper array). (dunno about lua, nor different versions of lua, but in many languages the variable goes out of scope and is invalid after the for loop)

also, we'll need to check that the surface actually exists in game.surfaces. 

(this idea seems better:)
alternatively, while we loop over the surfaces, we can just skip a surface if it's not in the array of desired surfaces (inside of _reinit_reentrant_level1_body where it checks if the surface is valid)? that might work better (so that we don't need to init/reinit maybe?)

NocturnalScream replied:
> surf here will always be just a single surface? need to have it build an array instead? also, it's sort of bad form to declare a local inside a for loop anyway (need to declare it outside to make it a proper array). (dunno about lua, nor different versions of lua, but in many languages the variable goes out of scope and is invalid after the for loop)

also, we'll need to check that the surface actually exists in game.surfaces.

Done

> (this idea seems better:)
alternatively, while we loop over the surfaces, we can just skip a surface if it's not in the array of desired surfaces (inside of _reinit_reentrant_level1_body where it checks if the surface is valid)? that might work better (so that we don't need to init/reinit maybe?)

A bit hesitant here - first I'm struggling a bit iterate through the surfaces - for some reason i mess that up every time I try. Secondly, isn't it cheaper to never iterate through all the game surfaces but only through the ones we specify?

Also we should add a check for when no valid surfaces are supplied to just use nauvis -> or print a message that Concreep is disabled till a valid surface is entered

MarximusMaximus replied:
> A bit hesitant here - first I'm struggling a bit iterate through the surfaces - for some reason i mess that up every time I try. Secondly, isn't it cheaper to never iterate through all the game surfaces but only through the ones we specify?

the reason to iterate and filter later is in case a surface gets created after initialization of our settings

> Also we should add a check for when no valid surfaces are supplied to just use nauvis -> or print a message that Concreep is disabled till a valid surface is entered

hmmm yea, probably print a message in ... init if the trimmed version of surfaces allowed is empty string

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -599,16 +608,31 @@ function is_valid_roboport_tile_names()
 	end
 end
 
+function runtime_settings_changed()
```
if we refer to `settings.global[<var>].value` for every usage, we shouldn't need most of this function. we can move all the script.on_nth_tick calls inside init.

NocturnalScream replied:
> if we refer to settings.global[<var>].value for every usage, we shouldn't need most of this function

Done.

> we can move all the script.on_nth_tick calls inside init.

also done, I just keep the function now to call init - also in case we might need it in the future for something else?
alternatively we can just call init directly when 
`script.on_event(defines.events.on_runtime_mod_setting_changed)` 
happens

MarximusMaximus replied:
hmmm, yea, sure, let's keep the function and for now it just calls init. it's an extra function call, but that's nbd here.

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -599,16 +608,31 @@ function is_valid_roboport_tile_names()
 		defines.events.on_entity_cloned, 
 		defines.events.script_raised_built,
-		defines.events.script_raised_revive
+		defines.events.script_raised_revive		
```
stray whitespace?

NocturnalScream replied:
Killed them all :)

MarximusMaximus commented on concreep-refilled/locale/en/locale.cfg:
```diff
@@ -7,9 +7,15 @@ creep refined concrete=Placer refined concrete
 cover landfill=Cover landfill
+minimum robot=Minimum amount of free robots in net
+minimum item=Minimum amount of items in logistic net to start concreep
+RUN_EVERY_N_UPDATES=Creeper-Update interval
```
since these will no longer be constants, let's not do all caps for these settings/localization strings

NocturnalScream replied:
Done

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -599,16 +608,31 @@ function is_valid_roboport_tile_names()
 	}, 
 	on_built_entity_handler
 )
 script.on_nth_tick(RUN_EVERY_N_UPDATES, creepers_update)
```
this line could get moved to init

NocturnalScream replied:
Done

MarximusMaximus commented on concreep-refilled/locale/en/locale.cfg:
```diff
@@ -7,9 +7,15 @@ creep refined concrete=Placer refined concrete
 debug=Various debug output (very spammy)
 debug function calls=Function call debug output (very spammy)
 debug coroutine calls=Coroutine call debug output (very spammy)
+surface concreep is running on= Specify surface concreep should be running on
```
suggested setting name: "allowed surfaces"
suggested localization: "Surfaces to creep on"

MarximusMaximus commented on concreep-refilled/locale/en/locale.cfg:
```diff
@@ -20,6 +26,12 @@ creep refined concrete=Whether or not to allow placing refined concrete. (Defaul
+minimum item= Concreep will only run if there is this amount of concrete/refined concrete etc. available. (Default: 200)
+RUN_EVERY_N_UPDATES=Changes in what interval Concreep is checking for possible creeper updates. Higher values use more UPS. (Default: 60)
+CREEPERS_PER_UPDATE=Changes how many Creepers are checked per update cycle. Higher values use more UPS. (Default: 5)
+ROBOPORTS_PER_REINIT_PASS=Changes how many Roboports are checked once concreep was inactive for more than 1 second. (Defualt: 1)
```
"default" is spelled wrong. maybe we should default this to something higher?

NocturnalScream replied:
Fixed the spelling.
Also fixed some localization being wrong :"for more than 1 second." into "more than 60 update cycles."
Also - sure we could do that. I just kept it at what it was. something like 3 work? however we don't want the game to freeze or stutter when a reinit is happening

MarximusMaximus replied:
Do you have a junkier machine than your main gaming rig? If so, could maybe test 1 vs 2 vs 3 vs 5 vs 10? I don't think my lesser machine is "lesser" enough, but I will try.

MarximusMaximus commented on concreep-refilled/locale/en/locale.cfg:
```diff
@@ -20,6 +26,12 @@ creep refined concrete=Whether or not to allow placing refined concrete. (Defaul
 debug=Lots of debug output for determining creeper state. (Default: false)
 debug function calls=Lots of output for determining events. (Default: false)
 debug coroutine calls=Lots of output for watching coroutine execution. (Default: false)
+surface concreep is running on= Intended to prevent concreep from running on all surfaces - Specify multiply surfaces with their names, i.e. Surface1,Surface2,... (Default: nauvis)
```
"Intended to prevent concreep from running on all surfaces" -> "Only allow concreep to affect specified surfaces"
"Specify multiply surfaces with their names" -> "specify multiple surfaces by their names separated by commas"

NocturnalScream replied:
> "Intended to prevent concreep from running on all surfaces" -> "Only allow concreep to affect specified surfaces"

Agreed.

> "Specify multiply surfaces with their names" -> "specify multiple surfaces by their names separated by commas"

not sure if this is better -> I think an example on how the user should input it will still be usefull

MarximusMaximus replied:
oh yes, keep the example too, i just wanted to alter the phrasing to include the part with needing a comma between them

MarximusMaximus commented on concreep-refilled/settings.lua:
```diff
@@ -78,4 +123,13 @@ data:extend({
 		default_value = false,
 		order = "992"
     },
+	{
```
let's move this setting to be above "minimum robot" setting, or at least above the debug settings

NocturnalScream replied:
Decided to move it to the top - in case the user is wondering why concreep doesn't work  on another surface -> first thing he sees in the settings will be this

MarximusMaximus replied:
aye. i was debating that or not. lol

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -306,24 +316,26 @@ function creep(creeper)
 	local brick_count = math.max(0, roboport.logistic_network.get_item_count("stone-brick") - MINIMUM_ITEM_COUNT)
 	local landfill_count = math.max(0, roboport.logistic_network.get_item_count("landfill") - MINIMUM_ITEM_COUNT)
-
+	local available_worker_count = math.max(0, roboport.logistic_network.available_construction_robots - MINIMUM_ROBOTS)
```
add this var to the debug print below it

NocturnalScream replied:
Done (I guess? Hope thats what you meant)

MarximusMaximus reviewed (changes requested):
Lots of mostly little tweak suggestions that should make it run better. Somewhat bigger change suggested for init and looping over surfaces.

NocturnalScream commented on concreep-refilled/locale/en/locale.cfg:
```diff
@@ -20,6 +26,12 @@ creep refined concrete=Whether or not to allow placing refined concrete. (Defaul
 cover landfill=Whether or not to concreep onto landfill. (Default: true)
+minimum robot= Concreep will only run if there is atleast this amount of free worker robots in the net available. (Default: 30)
+minimum item= Concreep will only run if there is this amount of concrete/refined concrete etc. available. (Default: 200)
+RUN_EVERY_N_UPDATES=Changes in what interval Concreep is checking for possible creeper updates. Higher values use more UPS. (Default: 60)
```
Localization error: Not higher values of Updates interval use more ups but lower.

NocturnalScream replied:
Well - fixed it as well - Didnt know i can't resolve a conversation without you - makes sense though :D so I guess -
resolved?

NocturnalScream committed:
requested Changes

Done most of the requested Changes:
- Renamed default vars to DEFAULT_VAR
- Using reference to settings.global for configuarable vars
Fixed whitespaces and incosnstitent formatting
- Removed the surface list building block from init()
Moved script.on.nth.tick unregister and register to init()
- cleared most of runtime_settings_changed() as we dont need it anymore
- fixed array building of the surfaces in _reinit_reentrant_level1_start() and added fail safe if no valid surfaces are specified (defaults to nauvis)
- no more all caps for non constants in settings/localization strings
- localization fixed
- changed order of settings for better accessibility

NocturnalScream reviewed (commented):
Fixed most of what was requested - have a few questions here and there - also we could think of adding a few more features if you agree (random radius, checks for specific tiles when landfilling, new tile palette system and change radius increments when no landfill is selected?)

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -64,14 +64,8 @@ function init()
-		end
-	end
+	script.on_nth_tick(nil) --unregister all nth tick handlers, so we can update the interval n
+	script.on_nth_tick(run every n updates, creepers_update)
```
missing settings.global ?

NocturnalScream replied:
whops - ye must have not used replace all in this instance.

NocturnalScream replied:
fixed

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -1,17 +1,17 @@
+DEFAULT_INITIAL_RADIUS = 1
 
-DEBUG_PREFIX = "Concreep: "
+DEFAULT_DEBUG_PREFIX = "Concreep: "
```
this one can stay in this file and doesn't need DEFAULT_ prefix, it's just a regular constant

NocturnalScream replied:
Done

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -55,7 +55,7 @@ function init()
 	global.creepers = {}
 	global.creeper_index = 1
-	global.robots_given_instructions = 0
+	global.robots_given_instructions = 0 --not used atm - what was the intention here?
```
was planning to remove this after merging your PR

NocturnalScream replied:
Well - leaving it up to you in that case - maybe we have some idea till then

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -121,7 +115,7 @@ function _reinit_reentrant()
 		-- no loops running, check if we should start outer loop
 		global.no_work_counter = global.no_work_counter + 1
-		if global.no_work_counter >= NO_WORK_UPDATES_BEFORE_REINIT then
+		if global.no_work_counter >= DEFAULT_NO_WORK_UPDATES_BEFORE_REINIT then
```
should get value from settings.global here

NocturnalScream replied:
done - sort of? I implemented them as hidden settings int the settings .lua as I thought we might make it user accessible at one point.

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -159,8 +165,8 @@ function _reinit_reentrant_level1_body()
-		-- -- limit how many surfaces checked per call
-		-- for pass_index = 1, SURFACES_PER_REINIT_PASS do
+		-- -- limit how many settings.global["allowed surfaces"].value checked per call
+		-- for pass_index = 1, DEFAULT_SURFACES_PER_REINIT_PASS do
```
should reference the correct settings.global here

NocturnalScream replied:
I definitely mixed something up with the replace all option today.. sigh thanks for pointing out! :)

NocturnalScream replied:
done

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -171,15 +177,15 @@ function _reinit_reentrant_level1_body()
-			--  - we either exhausted the list of surfaces (so return true) or 
-			--  - we reached SURFACES_PER_REINIT_PASS for this call (so return false)
+			--  - we either exhausted the list of settings.global["allowed surfaces"].value (so return true) or 
+			--  - we reached DEFAULT_SURFACES_PER_REINIT_PASS for this call (so return false)
```
should reference the correct settings.global here

NocturnalScream replied:
done

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -210,10 +216,9 @@ end
 	-- psuedo: for global.reinit_current_surface_roboport_index, #global.reinit_current_surface_roboports do
 		-- limit how many roboports checked per call
-		for pass_index = 1, ROBOPORTS_PER_REINIT_PASS do
+		for pass_index = 1, roboports per reinit pass do
```
missing "settings.global[]"

NocturnalScream replied:
done

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -230,7 +235,7 @@ function _reinit_reentrant_level2_body()
 		-- if we got here:
 		--  - we either exhausted the list of roboports (so return true) or
-		--  - we reached ROBOPORTS_PER_REINIT_PASS (so return false)
+		--  - we reached roboports per reinit pass (so return false)
```
reference that it's `settings.global["roboports per reinit pass"]`

NocturnalScream replied:
done

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -254,11 +259,11 @@ function creepers_update()
 		return 
 	end
-	for i = 1, CREEPERS_PER_UPDATE do
+	for i = 1, creepers per update do
```
missing settings.global[]

NocturnalScream replied:
fixed

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -525,7 +531,7 @@ function creep(creeper)
 		else
 			if #upgrade_target_types > 0 and surface.count_tiles_filtered{name=upgrade_target_types, area=area, limit=1} > 0 then
-				creeper.radius = INITIAL_RADIUS --Reset radius and switch to upgrade mode.
+				creeper.radius = DEFAULT_INITIAL_RADIUS --Reset radius and switch to upgrade mode.
```
use the settings.global here

NocturnalScream replied:
dixed

MarximusMaximus commented on concreep-refilled/control.lua:
```diff
@@ -584,12 +589,11 @@ function add_creeper_for_roboport(roboport)
 	local surface = roboport.surface
 
-	table.insert(global.creepers, {roboport = roboport, radius = INITIAL_RADIUS})
+	table.insert(global.creepers, {roboport = roboport, radius = DEFAULT_INITIAL_RADIUS})
```
use the settings.global here

NocturnalScream replied:
done

MarximusMaximus commented on concreep-refilled/settings.lua:
```diff
@@ -1,49 +1,58 @@
 data:extend({
```
we should move the new DEFAULT_s above here, and then reference them throughout the data struct below. makes maintenance a bit easier.

NocturnalScream replied:
check

NocturnalScream reviewed (commented):
-fixed missing/wrong settings.global references
-moved DEFAULT settings to settings.lua

NocturnalScream committed:
settings.global fixes for DEFAULTS

-fixed missing/wrong settings.global references
-moved DEFAULT settings to settings.lua

NocturnalScream commented on concreep-refilled/control.lua:
```diff
@@ -589,7 +576,7 @@ function add_creeper_for_roboport(roboport)
 
-	table.insert(global.creepers, {roboport = roboport, radius = DEFAULT_INITIAL_RADIUS})
+	table.insert(global.creepers, {roboport = roboport, radius = settings.global["DEFAULT_INITIAL_RADIUS"].value})
 end
```
- needs fixing to settings.global["inital radius"

NocturnalScream commented on concreep-refilled/control.lua:
```diff
@@ -165,8 +152,8 @@ function _reinit_reentrant_level1_body()
 	debug_print_coroutine_was_called("_reinit_reentrant_level1_body()")
```
line 132:  if name and game.surfaces[name].valid	-> then missing

NocturnalScream replied:
line 136 `if listOfSurfaces == {}` -> then missing

NocturnalScream commented on concreep-refilled/settings.lua:
```diff
@@ -1,4 +1,38 @@
+		type = "int-setting",
+		name = "initial radius",
+		setting_type = "runtime-global",
+		default_value = DEFAULT_INITIAL_RADIUS1,
```
- wrong variable name - accidental "1" slipped in here

NocturnalScream commented on concreep-refilled/control.lua:
```diff
@@ -235,7 +222,7 @@ function _reinit_reentrant_level2_body()
 		-- if we got here:
 		--  - we either exhausted the list of roboports (so return true) or
-		--  - we reached roboports per reinit pass (so return false)
+		--  - we reached settings.global["roboports per reinit pass"] (so return false)
```
`settings.global["roboports per reinit pass"]` .value is missing here

NocturnalScream commented on concreep-refilled/control.lua:
```diff
@@ -185,7 +172,7 @@ function _reinit_reentrant_level1_body()
 			-- if we got here:
 			--  - we either exhausted the list of settings.global["allowed surfaces"].value (so return true) or 
-			--  - we reached DEFAULT_SURFACES_PER_REINIT_PASS for this call (so return false)
+			--  - we reached settings.global["surfaces per reinit pass"] for this call (so return false)
```
.value missing

NocturnalScream commented on concreep-refilled/control.lua:
```diff
@@ -218,7 +205,7 @@ function _reinit_reentrant_level2_body()
 	-- psuedo: for global.reinit_current_surface_roboport_index, #global.reinit_current_surface_roboports do
 		-- limit how many roboports checked per call
-		for pass_index = 1, roboports per reinit pass do
+		for pass_index = 1, settings.global["roboports per reinit pass"] do
```
.value missing

NocturnalScream commented:
would close this at this point and instead focus on pr#3

MarximusMaximus committed:
testing modifying the PR

testing modifying the PR

MarximusMaximus commented:
closing as #3 supercedes it
